sql: Include RLS policies in EXPLAIN output

spilchen opened the pull request:
This change updates EXPLAIN output to list the RLS policies applied during query execution. Policies are shown only when the verbose option is used (e.g., `EXPLAIN (VERBOSE) …`) and at least one table with row-level security (RLS) is involved. The policy information is stored inline within the scan node.

Example output for a simple query:

```
                      info
------------------------------------------------
  distribution: local
  vectorized: true

  • scan
    columns: (c1)
    estimated row count: 1,000 (missing stats)
    table: t1@t1_pkey
    spans: FULL SCAN
    policies: pol1, pol2

(10 rows)
```

Epic: CRDB-11724
Release note: None

Closes #138824

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/141127)
<!-- Reviewable:end -->

spilchen commented:
Some background on why the policy information was added to the scan node: https://cockroachlabs.slack.com/archives/C083W9NK34H/p1737039471138559

michae2 reviewed (commented):
This looks pretty nice!

> Reviewed 15 of 15 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141127#-:-OJ0nuMo4GrUYe9Sw38-:b4aunmr)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`pkg/sql/opt/exec/execbuilder/relational.go` line 462 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141127#-OJ0cj14DniMALXRntjE:-OJ0cj14DniMALXRntjF:bqamcpt) ([raw file](https://github.com/cockroachdb/cockroach/blob/c64f81e00be605b4ab1f981d4c01e87315aaaeb7/pkg/sql/opt/exec/execbuilder/relational.go#L462)):*
> ```Go
> 			// value as this can be emitted when scanning a table and RLS forced all
> 			// rows to be filtered out because none of the policies applied.
> 			if len(values.Rows) == 0 && rlsMeta.NoPoliciesApplied {
> ```

Do we know for sure this ValuesExpr corresponds to a table scan that no RLS policies applied to? SimplifyZeroCardinalityGroup adds empty values expressions whenever it finds a zero-cardinality group. I don't think it's necessarily because of RLS.

Maybe we could somehow save the ValuesPrivate.ID in the metadata.
___
*[`pkg/sql/opt/exec/execbuilder/relational.go` line 465 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141127#-OJ0aky55S3ZZUtCvj8n:-OJ0aCqCBkriHndOJbo7:b-sv4ny1) ([raw file](https://github.com/cockroachdb/cockroach/blob/c64f81e00be605b4ab1f981d4c01e87315aaaeb7/pkg/sql/opt/exec/execbuilder/relational.go#L465)):*
> ```Go
> 				ef.AnnotateNode(node, exec.PolicyInfoID, &exec.RLSPoliciesApplied{})
> 			}
> 		} else if scan, ok := e.(*memo.ScanExpr); ok {
> ```

There are other operators that can read directly from a table (e.g. LookupJoinExpr, ZigzagJoinExpr, InvertedJoinExpr, PlaceholderScanExpr, IndexJoinExpr, VectorSearchExpr). After optimization, I think RLS could also apply to those operators.
___
*[`pkg/sql/opt/exec/explain/flags.go` line 77 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141127#-OJ0oChV2o5o4lXO4N5c:-OJ0oChV2o5o4lXO4N5d:brpsjcp) ([raw file](https://github.com/cockroachdb/cockroach/blob/c64f81e00be605b4ab1f981d4c01e87315aaaeb7/pkg/sql/opt/exec/explain/flags.go#L77)):*
> ```Go
> 	if options.Flags[tree.ExplainFlagTypes] {
> 		f.Verbose = true
> 		f.ShowTypes = true
> ```

Usually we consider "TYPES" to be more verbose than "VERBOSE" so I'd add it here as well. (Or use the Verbose flag instead of creating a new flag.)


<!-- Sent from Reviewable.io -->

spilchen committed:
sql: Include RLS policies in EXPLAIN output

This change updates EXPLAIN output to list the RLS policies applied during
query execution. Policies are shown only when the verbose option is used (e.g.,
`EXPLAIN (VERBOSE) …`) and at least one table with row-level security (RLS) is
involved. The policy information is stored inline within the scan node.

Example output for a simple query:

```
                      info
------------------------------------------------
  distribution: local
  vectorized: true

  • scan
    columns: (c1)
    estimated row count: 1,000 (missing stats)
    table: t1@t1_pkey
    spans: FULL SCAN
    policies: pol1, pol2

(10 rows)
```

Epic: CRDB-11724
Release note: None

Closes #138824

spilchen reviewed (commented):
Thanks for taking a look. I addressed two of your comments. I could use your help with the comment on the `ValuesExpr`.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141127#-:-OJ3f1_a2veEwpp2jwQi:b-fjplzb)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @michae2)

___
*[`pkg/sql/opt/exec/execbuilder/relational.go` line 462 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141127#-OJ0cj14DniMALXRntjE:-OJ3Y1SR72wtDpSV21JH:b-j0aged) ([raw file](https://github.com/cockroachdb/cockroach/blob/c64f81e00be605b4ab1f981d4c01e87315aaaeb7/pkg/sql/opt/exec/execbuilder/relational.go#L462)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Do we know for sure this ValuesExpr corresponds to a table scan that no RLS policies applied to? SimplifyZeroCardinalityGroup adds empty values expressions whenever it finds a zero-cardinality group. I don't think it's necessarily because of RLS.

Maybe we could somehow save the ValuesPrivate.ID in the metadata.
</blockquote></details>

I don't set up the `ValuesExpr` directly. Instead, I add a filter to a select operator. If no RLS policies apply, I use the `memo.FalseSingleton` filter to ensure that no rows are returned. This filter is later converted into a `ValuesExpr`, but I can't easily determine what its `ValuesPrivate.ID` will be. Whenever I use `memo.FalseSingleton`, I also set `NoPoliciesApplied`. Should I explicitly create a `ValuesExpr` myself to ensure I know the ID? That seems like the optimizer's responsibility, though. Or do you think checking `NoPoliciesApplied` is good enough?
___
*[`pkg/sql/opt/exec/execbuilder/relational.go` line 465 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141127#-OJ0aky55S3ZZUtCvj8n:-OJ3bTUqDtd5PG5TWKyr:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/c64f81e00be605b4ab1f981d4c01e87315aaaeb7/pkg/sql/opt/exec/execbuilder/relational.go#L465)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

There are other operators that can read directly from a table (e.g. LookupJoinExpr, ZigzagJoinExpr, InvertedJoinExpr, PlaceholderScanExpr, IndexJoinExpr, VectorSearchExpr). After optimization, I think RLS could also apply to those operators.
</blockquote></details>

Done
___
*[`pkg/sql/opt/exec/explain/flags.go` line 77 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141127#-OJ0oChV2o5o4lXO4N5c:-OJ3byCZ5kjaFf6CvVlN:b-4frmfk) ([raw file](https://github.com/cockroachdb/cockroach/blob/c64f81e00be605b4ab1f981d4c01e87315aaaeb7/pkg/sql/opt/exec/explain/flags.go#L77)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Usually we consider "TYPES" to be more verbose than "VERBOSE" so I'd add it here as well. (Or use the Verbose flag instead of creating a new flag.)
</blockquote></details>

Thanks. I added the setting for `f.ShowPolicyInfo` separately because I didn't want to tie policy information directly to the verbose flag. In the explain test I added, I wanted to see the policy information without enabling all verbose details.


<!-- Sent from Reviewable.io -->

michae2 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> I'm giving up on the Values thing, it seems too difficult. Nice job!

> Reviewed 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141127#-:-OJ4WWbWB8Zecp_26llm:b-vocjs2)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`pkg/sql/opt/exec/execbuilder/relational.go` line 462 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141127#-OJ0cj14DniMALXRntjE:-OJ4TZiqA-_UEsg8kXvc:biov1bi) ([raw file](https://github.com/cockroachdb/cockroach/blob/c64f81e00be605b4ab1f981d4c01e87315aaaeb7/pkg/sql/opt/exec/execbuilder/relational.go#L462)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

I don't set up the `ValuesExpr` directly. Instead, I add a filter to a select operator. If no RLS policies apply, I use the `memo.FalseSingleton` filter to ensure that no rows are returned. This filter is later converted into a `ValuesExpr`, but I can't easily determine what its `ValuesPrivate.ID` will be. Whenever I use `memo.FalseSingleton`, I also set `NoPoliciesApplied`. Should I explicitly create a `ValuesExpr` myself to ensure I know the ID? That seems like the optimizer's responsibility, though. Or do you think checking `NoPoliciesApplied` is good enough?
</blockquote></details>

I see, that makes sense.

Let's leave it the way you have it. I don't think there's an easy way to get the `ValuesPrivate.ID` (and it might change, anyway, if the optimizer creates a new empty Values higher in the tree). We might annotate the wrong `norows` here, but that seems better than not explaining why the table scan disappeared from the plan.


<!-- Sent from Reviewable.io -->

spilchen commented:
TFTR @michae2 

bors r+
